consensus/hotstuff/verification/combined_verifier_v2.go

```diff
@@ -97,9 +104,17 @@ func (c *CombinedVerifier) VerifyVote(signer *flow.Identity, sigData []byte, blo
 	return nil
 }
 
-// VerifyQC verifies the validity of a combined signature on a quorum certificate.
+// VerifyQC checks the cryptographic validity of the QC's `sigData` for the
+// given block. It is the responsibility of the calling code to ensure
+// that all `voters` are authorized, without duplicates. Return values:
+//  - nil if `sigData` is cryptographically valid
+//  - model.ErrInvalidFormat if `sigData` has an incompatible format
+//  - model.ErrInvalidSignature if a signature is invalid
+//  - error if running into any unexpected exception (i.e. fatal error)
 func (c *CombinedVerifier) VerifyQC(signers flow.IdentityList, sigData []byte, block *model.Block) error {
-
+	if len(signers) == 0 {
```

Comment:
Good question. Thanks. I'll document the answer in the code as well. 
Reasoning:
* The function call to [`AggregateBLSPublicKeys`](https://github.com/onflow/flow-go/blob/e61f2934f815e1877cd616965d36753df26b249b/crypto/bls_multisig.go#L159-L171), which we do in line 142, https://github.com/onflow/flow-go/blob/e61f2934f815e1877cd616965d36753df26b249b/consensus/hotstuff/verification/combined_verifier_v2.go#L142-L145 returns a `crypto.invalidInputsError` in two distinct cases:
   1. In case no keys are provided, i.e.  `len(signers) == 0`. 
      This scenario _is expected_ during normal operations, because a byzantine proposer might construct an (invalid) QC with an empty list of signers. 
   2. In case some of the provided public keys are invalid. 
       This scenario is _not expected_ during normal operations, because all keys are supposed to be checked _before_ they enter the protocol state bases on a proof of possession.

In case of (i), we have an invalid QC. In case of (ii), we have a bug (or uncovered edge case) in our implementation. At the moment, we don't have a proof of possession, i.e. there is the possibility of  having invalid keys in the protocol state. 

The check `len(signers) == 0` is for robustness of the code: 
* By checking `len(signers) == 0` upfront, we can rule out case (i) as a source of error. Hence, if we encounter an error in line 142, we know it is case (ii). 
* Thereby, we can clearly distinguish a faulty _external_ input from an _internal_ bug.

Reply:
I added the following documentation to the code: https://github.com/onflow/flow-go/blob/44d665742ac284822af7233fc564a766a6837d33/consensus/hotstuff/verification/combined_verifier_v2.go#L142-L157